design/DFX-Interface.md

```diff
@@ -95,23 +96,24 @@ Exporting Canister Metadata
 ---------------------------
 
 The compiler generates various metadata about the canister via command line flags.
-The compiled Wasm module also includes the metadata in the custom sections.
+The compiled Wasm module also includes these metadata in the custom sections.
+The compiler flag `--public-metadata <name>` controls if the custom section is publicly accessible.
+If `<name>` is in the `public-metadata` flag, the custom section name will be `icp:public <name>`.
+Otherwise, it will be `icp:private <name>`.
 
 * Candid interface.
   + Compiler flag `--idl` generates the Candid interface for the canister. The main service
     is always a service constructor, which contains the initialization arguments for installing the canister.
-  + Custom section `icp:public candid` stores the interface for the running (initialized) canister, which removes
-    the initializatio arguments. This custom section is publicly accessible.
-  + Custom section `icp:private init_args` stores the initialization arguments. The argument types can refer to
-    types defined in the `icp:public candid` custom section. This custom section is only accessible by the controllers
-    of the canister.
+  + Custom section `candid:service` stores the interface for the running (initialized) canister, which removes
+    the initializatio arguments.
+  + Custom section `candid:args` stores the initialization arguments. The argument types can refer to
+    types defined in the `candid:service` custom section.
```

Comment:
From the candid perspective, service constructor is better, because it allows you to specify higher-order actor class (although we never support it).

I slightly prefer the current `--idl` output. Because `--idl` outputs the type of the compiled Wasm that is yet to be installed. For consistency, even `actor {}` will have the type `() -> service {}`. The custom section stores the interface for the running canister, so it's always `service {}` without the init args. Without the first class support for actor class, the `candid:init-args` section is not even needed.

Another minor point is that the output file is supposed to be self-contained. That means `--init-arg-types <file>` will need to duplicate some of the type definitions from the did file. In the custom section, init args can refer to types from the main service, because arguably it's in the same Wasm module.

Reply:
It is specified: 
```
The argument types can refer to types defined in the `candid:service` custom section.
```

Reply:
Sorry, my bad, didn't see that. Too much reviewing on the phone.